Update to React 17 and latest WordPress packages

scinos opened the pull request:
## Changes proposed in this Pull Request
All updates required for React 17 and the latest WP monorepo packages.

### Testing instructions:
- [x] 1. Ensure that Calypso runs locally
- [x] 2. CI should pass
- [x] 3. Smoke test all portions of calypso
- [x] 4. Smoke test ETK
- [x] 5. Smoke test wpcom block editor
- [x] 6. Smoke test o2-blocks
- [x] 7. Test that the removal of custom `base-styles` works properly.
- [ ] 8. Double-check bundle sizes and m

ake sure those are ok.

## Preparation:
These tasks can be done in `trunk` in before merging React 17, as they will reduce the surface area of the React update. **_CI will fail in this PR until all prep tasks are merged and this PR is rebased_.**

### Assorted fixes:
- [x] #56004
- [x] #55971
- [x] #55997
- [x]  #53222
- [x] #56002 (ultimately completed in #53366)

### 3rd party dependency updates:
- [x] Removed `chai-enzyme` (#55931)
- [x] Updated Storybook (#54782)
- [x] Updated Emotion (#55736)
- [x] Updated Redux (#55999)
- [x] Updated Sass (#55304)
- [x] Updated react-resize-aware (#55929)
- [x] Updated `use-memo-one` (#55928)
- [x] Replace `react-lazily-render` (#56154)
- [x] #56089

### 1st party dependency updates:
- [x] `social-logos` (https://github.com/Automattic/social-logos/pull/109)
- [x] `gridicons` (https://github.com/Automattic/gridicons/pull/318)
- [x] `react-virtualized` (https://github.com/Automattic/react-virtualized/pull/2)
- [x] Integrate above changes to calypso: https://github.com/Automattic/wp-calypso/pull/55907
- [x] `newspack-blocks`. (https://github.com/Automattic/newspack-blocks/pull/855)
- [x] Update newspack blocks in calypso (https://github.com/Automattic/wp-calypso/pull/56121)
- [x] Updated `isolated-block-editor` (#56000)
- [ ] `newspack-components` (https://github.com/Automattic/newspack-plugin/pull/1135)

## Changes in this PR

### Dependency updates:
- Update to react 17: `react`, `react-dom`, `@types/react`
- All WP monorepo package updates from #54872

### Assorted changes:
- Use `@wojtekmaj/enzyme-adapter-react-17` for React 17, since the main enzyme adapter has not been updated.
- Update snapshot tests.
- Update all peer dependency requirements in `.yarnrc.yml`.
- Remove type hacks for `@wordpress/compose` now that it includes types.
- Add `global.CSS` mock for tests.
- Fix type changes.
- Include `jest-canvas-mock` in a test suite which should have required it already.
- Resolve `@types/react` to v17 for all dependencies to fix `tsc` compilation errors.
- Make sure that `jest-resolve` has a dependency on the correct version of `jest-haste-map`
- Remove resolutions which downgraded the wp base styles packages for compatibility.

### Ignored dependencies:
The following dependencies have peer deps on react 16, but have not received react 17 updates. They do not break anything, so we are ignoring the warning messages from them. Most are transitive dependencies from 3rd parties, such as storybook.

- `react-autosize-textarea`
- `airbnb-prop-types`
- `react-dates`
- `react-with-direction`
- `react-simple-code-editor`
- `create-react-context`
- `@reach/router`

scinos committed:
Update enzyme adapter

Use @wojtekmaj/enzyme-adapter-react-17

scinos committed:
Update snapshots

noahtallen commented:
There are a large number of sass deprecation warnings in the build log. Unfortunately, they won't be a quick fix. Since @wordpress/scripts updated their sass dependency, our yarn.lock is also updating it to a newer minor version, which I'm guessing includes the deprecation warnings.

They are easy to fix with the sass migrator tool, but this caused some other build problems. Specifically, the migrator tool adds `@use sass:math` to the top of files needing the math dependency. Interestingly, `@use` statements must be the first thing in every file -- and our compile step adds an `@import` prelude to every sass file. So we'll need to resolve that to get the build working with the new format. I think that would be nice to split out from this PR, but it means putting up with the deprecation warnings for the time being.

```
DEPRECATION WARNING: Using / for division is deprecated and will be removed in Dart Sass 2.0.0.

Recommendation: math.div($pixels, $context)

More info and automated migrator: https://sass-lang.com/d/slash-div
```

noahtallen committed:
Fixes for wp monorepo update

Squashed commit of the following:

commit 95ca82b
Author: Sergio Cinos <sergio.cinos@gmail.com>
Date:   Thu Jul 22 11:27:51 2021 +0200

    Add global CSS

commit 1e0c99d
Author: Sergio Cinos <sergio.cinos@gmail.com>
Date:   Thu Jul 22 09:36:36 2021 +0200

    Do not import types-patch

commit dee6bd6
Author: Sergio Cinos <sergio.cinos@gmail.com>
Date:   Thu Jul 22 09:29:26 2021 +0200

    Delete unused types patch

commit 2465d15
Author: Sergio Cinos <sergio.cinos@gmail.com>
Date:   Thu Jul 22 09:18:20 2021 +0200

    Fix types

noahtallen committed:
React 17 update

noahtallen committed:
Fix several peer dependency requirements

scinos commented on yarn.lock:
```diff
@@ -39358,22 +39351,22 @@ testarmada-magellan@11.0.10:
 typescript@^4.3.5:
-  version: 4.3.5
-  resolution: "typescript@npm:4.3.5"
+  version: 4.4.2
```
Was this intentional?

scinos commented:
> Some of these can be fixed by manually editing yarn.lock

I strongly suggest we avoid that solution: if a package's `package.json` and its entry in `yarn.lock` doesn't match anymore that's going to cause confusion in the future.

Alternative ideas:
* Hide the messages using yarn's `logFilters`.
* Fork the package and add react 17 as a peer dependency (assuming the package does work with react 17) + use `resolutions` to replace the upstream version with our fork. This has the benefit that can help the community, as most packages have open issues with a few votes from devs asking to include react 17 as a peer dep.

scinos committed:
Hide messages

scinos commented:
I hide the messages related to packages out of our control and that seem to be abandoned. After that, the list of missing dependencies is:

```
➤ YN0000: ┌ Resolution step
➤ YN0060: │ @automattic/calypso-stripe@workspace:packages/calypso-stripe provides react (p6c207) with version 17.0.2, which doesn't satisfy what react-stripe-elements requests
➤ YN0060: │ @automattic/calypso-stripe@workspace:packages/calypso-stripe provides react-dom (pebb3c) with version 17.0.2, which doesn't satisfy what react-stripe-elements requests
➤ YN0002: │ @automattic/wpcom-editing-toolkit@workspace:apps/editing-toolkit doesn't provide @wp-g2/create-styles (p2495c), requested by @automattic/newspack-blocks
➤ YN0060: │ @automattic/wpcom-editing-toolkit@workspace:apps/editing-toolkit provides react (p70a41) with version 17.0.2, which doesn't satisfy what @automattic/newspack-blocks and some of its descendants request
➤ YN0060: │ @automattic/wpcom-editing-toolkit@workspace:apps/editing-toolkit provides react-dom (p0b858) with version 17.0.2, which doesn't satisfy what @automattic/newspack-blocks and some of its descendants request
➤ YN0060: │ calypso@workspace:client provides react (p95a6a) with version 17.0.2, which doesn't satisfy what react-stripe-elements requests
➤ YN0060: │ calypso@workspace:client provides react (p141ab) with version 17.0.2, which doesn't satisfy what react-lazily-render requests
➤ YN0060: │ calypso@workspace:client provides react-dom (pbddb5) with version 17.0.2, which doesn't satisfy what react-stripe-elements requests
➤ YN0000: │ Some peer dependencies are incorrectly met; run yarn explain peer-requirements <hash> for details, where <hash> is the six-letter p-prefixed code
➤ YN0000: └ Completed in 0s 739ms
```

Those can/will be solved by:
- Update newspack blocks (newspack-blocks#855)
- Drop/fork react-lazily-render
- Migrate to react-stripe-elements to @stripe/react-stripe-js

scinos commented:
Error building the docker image:

```
15:57:09   #15 163.1 [evergreen] [webpack-cli] ModuleNotFoundError: Module not found: Error: Can't resolve 'process/browser' in '/calypso/node_modules/lib0'
15:57:09   #15 163.1 [evergreen] Did you mean 'browser.js'?
15:57:09   #15 163.1 [evergreen] BREAKING CHANGE: The request 'process/browser' failed to resolve only because it was resolved as fully specified
15:57:09   #15 163.1 [evergreen] (probably because the origin is strict EcmaScript Module, e. g. a module with javascript mimetype, a '*.mjs' file, or a '*.js' file where the package.json contains '"type": "module"').
15:57:09   #15 163.1 [evergreen] The extension in the request is mandatory for it to be fully specified.
15:57:09   #15 163.1 [evergreen] Add the extension to the request.
15:57:09   #15 163.1 [evergreen]     at /calypso/node_modules/webpack/lib/Compilation.js:1773:28
15:57:09   #15 163.1 [evergreen]     at /calypso/node_modules/webpack/lib/NormalModuleFactory.js:811:13
15:57:09   #15 163.1 [evergreen]     at eval (eval at create (/calypso/node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:10:1)
15:57:09   #15 163.1 [evergreen]     at /calypso/node_modules/webpack/lib/NormalModuleFactory.js:286:22
15:57:09   #15 163.1 [evergreen]     at eval (eval at create (/calypso/node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:9:1)
15:57:09   #15 163.1 [evergreen]     at /calypso/node_modules/webpack/lib/NormalModuleFactory.js:442:22
15:57:09   #15 163.1 [evergreen]     at /calypso/node_modules/webpack/lib/NormalModuleFactory.js:124:11
15:57:09   #15 163.1 [evergreen]     at /calypso/node_modules/webpack/lib/NormalModuleFactory.js:673:25
15:57:09   #15 163.1 [evergreen]     at /calypso/node_modules/webpack/lib/NormalModuleFactory.js:882:8
15:57:09   #15 163.1 [evergreen]     at /calypso/node_modules/webpack/lib/NormalModuleFactory.js:1002:5
```

I'll try to fix it.

noahtallen committed:
Fix typescript issues

- @types/react had to be pinned to v17 to prevent TSC compilation
  errors. This is because TSC would sometimes resolve types for
  react 16, which isn't used by us, but gets included by some
  dependencies.
- jest-haste-map had to be pinned to v27 because a different part
  of jest was importing from jest-haste-map without having had
  jest-haste-map in its package.json. As a result, it would resolve
  an older version of jest-haste-map, which had incorrect types.

scinos commented on packages/search/.storybook/main.js:
```diff
@@ -1,4 +1,9 @@
+// Note: `require.resolve` does not work because that will resolve a specific
+// file instead of the directory, meaning that "@emotion/styled/base" will not
+// resolve correctly.
+const toPath = ( module ) => resolve( __dirname, '../../../node_modules', module );
```
Would `path.dirname(require.resolve('@emotion/styled/package.json'))` work?

I'd like to avoid hardcoding paths to `node_modules` if possible, because it makes assumptions about package hoisting that may not be true in the future (I found lots of those in the yarn v3 update).

noahtallen replied:
I'll try that too

noahtallen replied:
Done in #55997

fullofcaffeine commented:
>  Fork, replace, or remove react-lazily-render

Yeah, it does seem to be unmaintained. The last npm publish date was ~2 years ago. It was introduced in [this PR](https://github.com/Automattic/wp-calypso/pull/43952) around a year ago. I'm tending towards forking and updating it to support React 17, while keeping the same API to avoid further changes in Calypso. I can experiment with it next week if no one beats me to it. Also 🏓 @olaseni in case he has any insights.

noahtallen commented:
Yeah, I actually had a conversation with @hambai, also involved in that domains work, and he mentioned that the domains page is meant to be used by people with lots of domains, so removing it would be a step backwards for that goal. I'm currently researching this library: https://www.npmjs.com/package/react-intersection-observer It has a lot of benefits:
- Well maintained
- Popular
- Doesn't add any extra transitive dependencies
- Might even be a semi-drop-in experience
- Uses intersection observer to handle this behavior in a more modern way

I'll spin up a PR to try that today :)

noahtallen committed:
Update yarn.lock

renovate-bot committed:
fix(deps): update wordpress monorepo

noahtallen committed:
More peer dependency fixes

noahtallen commented:
Alright, after the latest push, I see only these messages when running `yarn`:

```
➜  wp-calypso git:(update/react-17) ✗ yarn
➤ YN0000: ┌ Resolution step
➤ YN0060: │ @automattic/calypso-stripe@workspace:packages/calypso-stripe provides react (p6c207) with version 17.0.2, which doesn't satisfy what react-stripe-elements requests
➤ YN0060: │ @automattic/calypso-stripe@workspace:packages/calypso-stripe provides react-dom (pebb3c) with version 17.0.2, which doesn't satisfy what react-stripe-elements requests
➤ YN0060: │ @automattic/wpcom-editing-toolkit@workspace:apps/editing-toolkit provides react (pa4cd8) with version 17.0.2, which doesn't satisfy what @automattic/newspack-blocks and some of its descendants request
➤ YN0060: │ @automattic/wpcom-editing-toolkit@workspace:apps/editing-toolkit provides react-dom (pe0d22) with version 17.0.2, which doesn't satisfy what @automattic/newspack-blocks and some of its descendants request
➤ YN0060: │ calypso@workspace:client provides react (p95a6a) with version 17.0.2, which doesn't satisfy what react-stripe-elements requests
➤ YN0060: │ calypso@workspace:client provides react-dom (pbddb5) with version 17.0.2, which doesn't satisfy what react-stripe-elements requests
➤ YN0000: │ Some peer dependencies are incorrectly met; run yarn explain peer-requirements <hash> for details, where <hash> is the six-letter p-prefixed code
➤ YN0000: └ Completed in 0s 462ms
➤ YN0000: ┌ Fetch step
➤ YN0013: │ 3605 packages were already cached
➤ YN0000: └ Completed in 1s 760ms
➤ YN0000: ┌ Link step
➤ YN0000: └ Completed in 2s 183ms
➤ YN0000: Done with warnings in 4s 898ms
```

There are only two:
- [ ] React stripe elements, which will be replaced soon by https://github.com/Automattic/wp-calypso/pull/56089
- [ ] newspack-blocks. @dkoo said they'll look into my PR next week: https://github.com/Automattic/newspack-plugin/pull/1135. However, it may take longer since there are several out of date dependencies there, and I'm not sure if there are any conflicts.

noahtallen committed:
Remove base styles resolutions

noahtallen commented:
I'm marking this ready for review, as the remaining peer dependency issues should not prevent us from reviewing and testing the rest of the PR :)

scinos commented:
I only get the warnings about newspack. We could hide them until the newspack update is gone, as long as we are confident it won't actually break with React 17

```
➤ YN0060: │ @automattic/wpcom-editing-toolkit@workspace:apps/editing-toolkit provides react (pa4cd8) with version 17.0.2, which doesn't satisfy what @automattic/newspack-blocks and some of its descendants request
➤ YN0060: │ @automattic/wpcom-editing-toolkit@workspace:apps/editing-toolkit provides react-dom (pe0d22) with version 17.0.2, which doesn't satisfy what @automattic/newspack-blocks and some of its descendants request
➤ YN0000: │ Some peer dependencies are incorrectly met; run yarn explain peer-requirements <hash> for details, where <hash> is the six-letter p-prefixed code
```

Other than that, looks good to me (GitHub won't let me approve because I was the original author)

noahtallen commented:
I'll do some more testing of that, but the peer dependency warnings are all about react 17. I assume these are being replaced by the latest Gutenberg plugin code once the block lands on wpcom, so I'll verify that assumption before merging.

noahtallen commented:
I'm noticing several deprecated components warnings. We should resolve these as a follow-up task:

- `wp.data.select( 'core/editor' ).getBlocks` is deprecated since version 5.3. Please use `wp.data.select( 'core/block-editor' ).getBlocks` instead`
- `wp.nux is deprecated since version 5.4. Note: wp.components.Guide can be used to show a user guide`
- `Using Toolbar without label prop is deprecated since version 5.6. Please use ToolbarGroup component instead`
- `Using custom components as toolbar controls is deprecated since version 5.6. Please use ToolbarItem, ToolbarButton or ToolbarDropdownMenu components instead`
- `wp.data.select( 'core/editor' ).getSelectedBlock` is deprecated since version 5.3. Please use `wp.data.select( 'core/block-editor' ).getSelectedBlock`
- `wp.editor.RichTextToolbarButton is deprecated since version 5.3. Please use wp.blockEditor.RichTextToolbarButton instead`
- `Button isDefault prop is deprecated since version 5.4. Please use variant="secondary" instead`
-

noahtallen committed:
Update yarnrc

noahtallen commented:
Current yarn output is:

```sh
➜  wp-calypso git:(update/react-17) ✗ yarn
➤ YN0000: ┌ Resolution step
➤ YN0000: │ Some peer dependencies are incorrectly met; run yarn explain peer-requirements <hash> for details, where <hash> is the six-letter p-prefixed code
➤ YN0000: └ Completed in 0s 467ms
➤ YN0000: ┌ Fetch step
➤ YN0013: │ 3604 packages were already cached
➤ YN0000: └ Completed in 1s 783ms
➤ YN0000: ┌ Link step
➤ YN0000: └ Completed in 2s 425ms
➤ YN0000: Done with warnings in 4s 890ms
```

I think this is good, as we will be reminded to resolve and update the peer requirement messages we are discarding

noahtallen committed:
Fix ignore plugin syntax
